design/volume-group-snapshot.md

```diff
@@ -0,0 +1,605 @@
+   Grouping:   
+     Group A: PVC1 and PVC2  
+     Group B: PVC1 and PVC3  
+   ItemBlock: All objects-Pod1, Pod2, PVC1, PVC2, and PVC3, are collected into a single item block.  
+   VolumeGroupSnapshots:  
+     PVC1 and PVC2 (group A) point to the same VGS (VGS (group: A)).   
+     PVC3 (group B) point to a different VGS (VGS (group: B)).  
+```mermaid
+flowchart TD
+    subgraph ItemBlock
+        P1[Pod1]
+        P2[Pod2]
+        PVC1[PVC1 group: A, group: B]
+        PVC2[PVC2 group: A]
+        PVC3[PVC3 group: B]
+    end
+
+    %% Pod mount relationships
+    P1 -->|mounts| PVC1
+    P1 -->|mounts| PVC2
+    P2 -->|mounts| PVC1
+    P2 -->|mounts| PVC3
+
+    %% Grouping for Group A: PVC1 and PVC2 are grouped into VGS_A
+    PVC1 --- PVC2
+    PVC1 -- "Group A" --> VGS_A[VGS Group A]
+    PVC2 -- "Group A" --> VGS_A
+
+    %% Grouping for Group B: PVC3 grouped into VGS_B
+    PVC3 -- "Group B" --> VGS_B[VGS Group B]
+    
+```
+
+#### Updates to CSI PVC plugin:
+  - This is the plugin which creates VolumeSnapshots for the CSI PVCs if required.
+  - We might encounter the following cases:
+    - PVC has VGS label and VGS already exists:
+      - We can check if VGS exists by listing the VGS in the namespace using the backup UUID 
+        along with the label used for the VGS (we will be creating VGS with backup UUID and user specified VGS  group label) and see if the PVC is part of any of the existing VGS for the particular backup operation we are currently in.
+      - If not then create the VGS object with the required labels (in this case backup UUID and the user specified VGS group label) and also add the PVC names (that are included in the VGS) as 
+        annotations. (This will be useful for our VGS restore operation where we would need PVC resource identifiers).
+      - For non-datamover case:
+        - Once VGS is created, add the VGS as an additional item to be included in the backup.
+        - The VGS BIA plugin will add the related VS as additional items later.
+      - For datamover case:
+        - We need to bypass creation of VS but need to create DataUploads for the VS instance that got created due to VGS creation.
+        - Fetch the correct VS instances that got created via VGS and is related to the current PVC.
+        - Finally create DataUpload object for this VS/PVC pair.
+        - Add the DataUpload as an additional item.
+    - PVC has VGS label and VGS exists:
+      - In this case we do not want to re-create VGS object.
+      - The required VS/DU for the PVC already exists as VGS instance exists for the PVC and backup UUID.
+    - PVC does not possess VGS label:
+      - Create VS and follow the existing legacy workflow.
+
+
+#### Add a new VolumeGroupSnapshot(VGS) BIA plugin
+  - We need this plugin only for non-datamover case.
+  - For a particular instance of VGS, this plugin will wait for all the related VS instances to be in ready state.
```

Comment:
As per @blackpiglet, for CSI snapshot backup (WITHOUT data mover), we also use the async operation to wait for VS to get to `readyToUse`.
Therefore, probably, we can also delegate the VS/VSC handling to the existing code in PVC BIA where VS/VSC will be waited for the appearance of snapshot handle.